src/sentry/rules/history/preview.py

```diff
@@ -198,36 +200,67 @@ def get_fired_groups(
+    condition_activity: GroupActivityMap,
+    dataset_map: Dict[str, Dataset],
 ) -> GroupActivityMap:
     """
     Filters the activity to contain only groups that have the most events in the past 2 weeks.
 
     Since frequency conditions require one snuba query per groups, we need to limit the number groups we process.
     """
-    if not condition_activity:
-        return condition_activity
-    # TODO: Also check other datasets for top groups
-    groups = raw_query(
-        dataset=Dataset.Events,
-        start=start,
-        end=end,
-        filter_keys={"project_id": [project.id]},
-        conditions=[("group_id", "IN", list(condition_activity.keys()))],
-        aggregations=[("count", "group_id", "groupCount")],
-        groupby=["group_id"],
-        order_by="-groupCount",
-        seleted_columns=["group_id"],
-        limit=FREQUENCY_CONDITION_GROUP_LIMIT,
-    ).get("data", [])
-
-    top_groups = {group["group_id"] for group in groups}
+    datasets = {dataset_map.get(group) for group in condition_activity.keys()}
+
+    # queries each dataset for top x groups and then gets top x overall
+    groups = []
+    for dataset in datasets:
+        if dataset is None:
+            continue
+        group_ids = list(condition_activity.keys())
+        kwargs = {
+            "dataset": dataset,
+            "start": start,
+            "end": end,
+            "filter_keys": {"project_id": [project.id]},
+            "conditions": [("group_id", "IN", group_ids)],
+            "aggregations": [("count", "group_id", "groupCount")],
+            "groupby": ["group_id"],
+            "order_by": "-groupCount",
+            "seleted_columns": ["group_id"],
+            "limit": FREQUENCY_CONDITION_GROUP_LIMIT,
+        }
+        if dataset == Dataset.Transactions:
+            kwargs["conditions"] = [[["hasAny", ["group_ids", ["array", group_ids]]], "=", 1]]
+            kwargs["aggregations"].append(("arrayJoin", ["group_ids"], "group_id"))
+        groups.extend(raw_query(**kwargs).get("data", []))
+
+    k = lambda x: x["groupCount"]
+    sorted_groups = sorted(groups, key=k, reverse=True)
+
+    top_groups = {group["group_id"] for group in sorted_groups[:FREQUENCY_CONDITION_GROUP_LIMIT]}
     top_activity = {
         group: activity for group, activity in condition_activity.items() if group in top_groups
     }
     return top_activity
 
 
+def get_group_dataset(condition_activity: GroupActivityMap) -> Dict[str, Dataset]:
+    """
+    Returns a dict that maps each group to its dataset. Assumes each group is mapped to a single dataset.
+    If the dataset is not found/supported, it is mapped to None.
+    """
+    group_categories = Group.objects.filter(id__in=condition_activity.keys()).values_list(
+        "id", "type"
+    )
+    dataset_map = {
+        group[0]: GROUP_CATEGORY_TO_DATASET.get(GROUP_TYPE_TO_CATEGORY.get(GroupType(group[1])))
+        for group in group_categories
+    }
+    return dataset_map
```

Comment:
nit: can probably avoid the `dataset_map` assignment and return the dict literal.